src/storage/mvcc/reader.rs

```diff
@@ -194,34 +194,33 @@ impl MvccReader {
         Ok(Some((commit_ts, write)))
     }
 
-    fn check_lock(&mut self, key: &Key, mut ts: u64) -> Result<Option<u64>> {
+    fn check_lock(&mut self, key: &Key, mut ts: u64) -> Result<()> {
         if let Some(lock) = self.load_lock(key)? {
-            if lock.ts <= ts {
-                if ts == u64::MAX && key.raw()? == lock.primary {
-                    // when ts==u64::MAX(which means to get latest committed version for
-                    // primary key),and current key is the primary key, returns the latest
-                    // commit version's value
-                    ts = lock.ts - 1;
-                } else {
-                    // There is a pending lock. Client should wait or clean it.
-                    return Err(Error::KeyIsLocked {
-                        key: key.raw()?,
-                        primary: lock.primary,
-                        ts: lock.ts,
-                        ttl: lock.ttl,
-                    });
-                }
+            if lock.ts > ts ||
+                lock.lock_type == LockType::Lock ||
+                ts == u64::MAX && key.raw()? == lock.primary {
+                // ignore the lock:
+                // 1. when lock.ts > ts.
+                // 2. when lock's type is Lock(which means it won't change the value).
+                // 3. when ts==u64::MAX(which means to get latest committed version for
+                //    primary key),and current key is the primary key.
+                return Ok(())
             }
+            // There is a pending lock. Client should wait or clean it.
+            return Err(Error::KeyIsLocked {
+                key: key.raw()?,
+                primary: lock.primary,
+                ts: lock.ts,
+                ttl: lock.ttl,
+            })
         }
-        Ok(Some(ts))
+        Ok(())
     }
 
     pub fn get(&mut self, key: &Key, mut ts: u64) -> Result<Option<Value>> {
         // Check for locks that signal concurrent writes.
         match self.isolation_level {
-            IsolationLevel::SI => if let Some(new_ts) = self.check_lock(key, ts)? {
-                ts = new_ts;
-            },
+            IsolationLevel::SI => self.check_lock(key, ts)?,
```

Comment:
What if there are lots of `Rollback ` during  (lock.ts ~ ts)? The related issue is https://github.com/pingcap/tikv/pull/2871

Reply:
Then we just scan and ignore those rollbacks.
In the normal case, there is no lock, we need to scan those rollbacks anyway.
I think this optimization doesn't worth the complexity cost.